doc/source/f2py/buildtools/index.rst

```diff
@@ -0,0 +1,96 @@
+.. _f2py-bldsys:
+
+=======================
+F2PY and Build Systems
+=======================
+
+In this section we will cover the various popular build systems and their usage
+with ``f2py``.
+
+.. note::
+   **As of November 2021**
+
+   The default build system for ``F2PY`` has traditionally been the through the
+   enhanced ``numpy.distutils`` module. This module is based on ``distutils`` which
+   will be removed in ``Python 3.12.0`` in **October 2023**; ``setuptools`` does not
+   have support for Fortran or ``F2PY`` and it is unclear if it will be supported
+   in the future. Alternative methods are thus increasingly more important.
+
+
+Basic Concepts
+===============
+
+Building an extension module which includes Python and Fortran consists of:
+
+- Fortran source(s)
+- One or more generated files from ``f2py``
+
+  + A ``C`` wrapper file is always created
+  + Code with modules require an additional ``.f90`` wrapper
+
+- ``fortranobject.{c,h}``
+
+  + Distributed with ``numpy``
+  + Can be queried via ``python -c "import numpy.f2py; print(numpy.f2py.get_include())"``
+
+- NumPy headers
+
+  + Can be queried via ``python -c "import numpy; print(numpy.get_include())"``
+
+- Python libraries and development headers
+
+Broadly speaking there are three cases which arise when considering the outputs of ``f2py``:
+
+Fortran 77 programs
+   - Input file ``blah.f``
+   - Generates
+
+     + ``blahmodule.c``
+     + ``f2pywrappers.f``
+
+   When no ``COMMON`` blocks are present only a ``C`` wrapper file is generated.
+   Wrappers are also generated to rewrite assumed shape arrays as automatic
+   arrays.
+
+Fortran 90 programs
+   - Input file ``blah.f``
+   - Generates:
+
+     + ``blahmodule.c``
+     + ``blah-f2pywrappers2.f90``
+
+   The secondary wrapper is used to handle code which is subdivided into
+   modules. It rewrites assumed shape arrays as automatic arrays.
+
+Signature files
+   - Input file ``blah.pyf``
+   - Generates:
+
+     + ``blahmodule.c``
+     + ``blah-f2pywrappers2.f90`` (occasionally)
+     + ``f2pywrappers.f`` (occasionally)
+
+   Signature files ``.pyf`` do not signal their language standard via the file
+   extension, they may generate the F90 and F77 specific wrappers depending on
+   their contents; which shifts the burden of checking for generated files onto
+   the build system.
```

Comment:
Maybe this should say it's planned to be fixed and link to an issue related to that?

Reply:
To clarify, are you talking only about `.f` vs. `.f90` files, or also `.f08`, `.f95`,...?
I have only recently begun getting back into Fortran, but I took away from browsing online material that the latter is not really "best practice".

Reply:
@rgommers This is a bit of a tricky situation. The best we can do in terms of a fix is:
- Generate all the files anyway, there are only 3 possible outputs --> 1 `c` file, 1 `f77wrapper` file and 1 `f90wrapper`
  - This means a lot of files for some projects, but other than being kinda annoying in that sense is the best way to be deterministic about this
-   Unify the two kinds of wrappers
- Add a top-level comment which lists the files generated by the signature file

All of these would kind of bother projects which depend on the current behaviour.

I'll open an issue for this and link to it.

@bilderbuchi technically `f2py` only cares about `.f` or `.f77` which signals that it is in fixed form, otherwise any file ending (even `.blah`) will be treated as free-form Fortran. There are no `-std=` flags added by `f2py` on the basis of the file ending (which is `np.distutils`'s concern anyway). None of the file endings should be used, and support for these are not planned.

Reply:
> All of these would kind of bother projects which depend on the current behaviour.

Adding empty files that are okay to be left out, so the introduction is backwards-compatible, would be great.

> I'll open an issue for this and link to it.

Thanks!